namer/namer.cc

```diff
@@ -15,392 +18,730 @@
-            // Reverse order is intentional: that's how Ruby does it.
-            auto &arg = *it;
-            if (ast::isa_tree<ast::EmptyTree>(arg.get())) {
-                continue;
-            }
-            if (arg->isSelfReference()) {
-                dest->emplace_back(arg->deepCopy());
-                continue;
-            }
-            if (isValidAncestor(arg.get())) {
-                dest->emplace_back(arg->deepCopy());
-            } else {
-                if (auto e = ctx.beginError(arg->loc, core::errors::Namer::AncestorNotConstant)) {
-                    e.setHeader("`{}` must only contain constant literals", send->fun.data(ctx)->show(ctx));
-                }
-                arg = ast::MK::EmptyTree();
+        switch (send->fun._id) {
+            case core::Names::declareFinal()._id:
+            case core::Names::declareSealed()._id:
+            case core::Names::declareInterface()._id:
+            case core::Names::declareAbstract()._id: {
+                Modifier mod;
+                mod.kind = NameKind::Class;
+                mod.owner = klass;
+                mod.loc = send->loc;
+                mod.name = send->fun;
+                foundNames->addModifier(move(mod));
+                break;
             }
+            default:
+                break;
         }
     }
 
-    void aliasMethod(core::MutableContext ctx, core::Loc loc, core::SymbolRef owner, core::NameRef newName,
-                     core::SymbolRef method) {
-        core::SymbolRef alias = ctx.state.enterMethodSymbol(loc, owner, newName);
-        alias.data(ctx)->resultType = core::make_type<core::AliasType>(method);
-    }
-
-    core::SymbolRef methodOwner(core::MutableContext ctx) {
-        core::SymbolRef owner = ctx.owner.data(ctx)->enclosingClass(ctx);
-        if (owner == core::Symbols::root()) {
-            // Root methods end up going on object
-            owner = core::Symbols::Object();
+    FoundNameRef getOwner() {
+        if (ownerStack.empty()) {
+            return FoundNameRef::root();
         }
-        return owner;
+        return ownerStack.back();
     }
-    bool isValidAncestor(ast::Expression *exp) {
-        if (ast::isa_tree<ast::EmptyTree>(exp) || exp->isSelfReference() || ast::isa_tree<ast::ConstantLit>(exp)) {
-            return true;
-        }
-        if (auto lit = ast::cast_tree<ast::UnresolvedConstantLit>(exp)) {
-            return isValidAncestor(lit->scope.get());
+
+    // Returns index to foundNames containing the given name. Recursively inserts class refs for its owners.
+    FoundNameRef squashNames(core::Context ctx, const unique_ptr<ast::Expression> &node) {
+        if (auto *id = ast::cast_tree<ast::ConstantLit>(node.get())) {
+            // Already defined. Insert a foundname so we can reference it.
+            auto sym = id->symbol.data(ctx)->dealias(ctx);
+            ENFORCE(sym.exists());
+            return foundNames->addSymbol(sym);
```

Comment:
I debated encoding the symbolref in `FoundNameRef` directly, but I wasn't sure if it was safe to steal the top bits for the `NameKind` tag.